[FEATURE] Transactional editing for postgres layers

manisandro opened the pull request:
This is a re-worked patch of pull #1367. From #1367:

This pull request adds a transaction mode to a group of layers. Within these layers, edits are bypassing the QGIS edit buffer and are directly written into a database transaction. The edits are visible for layers outside the group or for external tools once the transaction is commited.

Further details in ticket #1367. 

Compared to the initial version in #1367, this patch does the following differently:
- `QgsTransaction` is an abstract class, each provider supporting transactions provides an appropriate concrete class.
- The `QgsPostgresTransaction` class holds the connection, the connection pool is not touched anymore.
- A "pass-through" edit buffer is introduced to forward layer changes directly to the data provider, avoiding lots of fragile if-else code in the `QgsVectorLayer`
- The postgres provider consistently uses the transaction connection, if in transaction mode, for all queries.

A small plugin to test this is here https://github.com/manisandro/qgis-transaction-test-plugin

manisandro committed:
[FEATURE]: Transactional editing for postgres layers (other datasources possible in the future). Funded by City of Uster

wonder-sk commented:
Nice - this looks much better than #1367 - and addresses various concerns I had before. Also thanks for the test plugin - it is possible to see transactions in action.

Testing:
- there are various scenarios with concurrency issues:
  - open two instances of QGIS - A and B. load the same layer, start transaction in both, start editing in both. In A change attribute value of a feature. In B try to delete the same feature. As a result, B will hang because Postgres is waiting for the other transaction to be committed/aborted. Is that acceptable? (Only once A commits / rolls back transaction, B will come back to life.)
  - open two instances again. In A delete feature, in B change its attribute (freeze). In A commit the transaction. In B (now unfrozen) commit the transaction - no error is given that the feature does not exist and the attribute cannot be changed.
  - open two instances again. In A modify row 1, in B modify row 2. In A modify row 2 (hangs). In B modify row 1. The change is aborted by Postgres to avoid deadlock, but no error is reported
- it is assumed that first a transaction is started, then editing is started, later stopped and finally transaction is committed. If the transaction is started/stopped while editing, the GUI is not working correctly

Code:
- in postgres feature source / feature iterator - if used in a worker thread (e.g. when rendering), transaction may get deleted in the meanwhile - likelihood of crashes!
- data provider's setTransaction() - please provide comments what it does and what are the ownership semantics. Also, I would prefer to have it as a virtual function which would have implementation only in providers supporting it (same for mTransaction). It should return true/false on success/failure. The postgres provider should also verify the instance has correct type.
- please add QgsTransaction class comments explaining how is it intended to be used, with implications on layers, getFeatures(), undo buffer etc.
- it is worth explaining in comments when/why a connection needs to be locked

[rant]The undo/redo system is not working at all - but obviously others do not see it as a big usability problem[/rant]

jef-n commented:
Hm, this is still just for plugins, right?   So probably no user will actually notice that it circumvents undo/redo.   To me it still doesn't feel right.

Hm, could there be a transaction/connection object that can be shared between multiple instances of a provider and the undo/redo stack tied to that instead of just one layer (ie. provider instance).

In the postgresql case (re)doing a command would create a savepoint and execute the corresponding database operations and undoing would restore the savepoint.

mhugent commented:
>Hm, this is still just for plugins

Yes, the normal user won't notice it

>Hm, could there be a transaction/connection object that can be shared between multiple instances of >a provider and the undo/redo stack tied to that instead of just one layer (ie. provider instance).

Yes, a transaction object can be shared between multiple instances and layers, so undo/redo with savepoints is a possibility. It is however difficult to guarantee that after doing undo and redo, the result is still the same as before. Having triggers, sequences, uuid generators, ... in the DB is very common in application modules (at least in QGEP0. Imho, not having an undo/redo stack is better than having one which might produce surprising results to the user.

manisandro committed:
Fix infinite loop in QgsVectorLayer::deleteSelectedFeatures if removal of feature fails

manisandro commented:
To fix the hangs, I now set a `statement_timeout`, currently hardcoded to 5000 milliseconds. Probably this is to short, and I guess it should be made configurable, for instance as an optional argument of `QgsPostgresTransaction::beginTransaction`, although it may be that for some providers the value will be ignored. Situations like 

> open two instances again. In A delete feature, in B change its attribute (freeze). In A commit the transaction. In B (now unfrozen) commit the transaction - no error is given that the feature does not exist and the attribute cannot be changed.

are now less likely to happen since the statement is aborted due to `statement_timeout`. Clearly, it can still happen that instance A commits the transaction before the timeout, in which case this would still happen. However, I don't think there is anything we can do about this on the QGIS side since Postgres does not seem to report any errors.

Layers can now only be added to a transaction if they are not editable, and editing needs to be disabled to finish the transaction. Actually, there is no hard reason to enforce this, since changes are written directly to the data provider anyway, but I guess it does not hurt to enforce this for UI consistency.

`transaction()` and `setTransaction()` made virtual as suggested, however I also made `setTransaction()` private and added a friendship relation with `QgsTransaction`. Not a big fan of the `friend` keyword, but in this case I think it makes sense since `setTransaction` really should only be called by `QgsTransaction`.

QgsTransaction documentation improved, added lots of sanity checks.

Comments added to `QgsPostgresConn::{lock/unlock}'

Open issue is how to prevent
> in postgres feature source / feature iterator - if used in a worker thread (e.g. when rendering), transaction may get deleted in the meanwhile - likelihood of crashes!

Any ideas?

wonder-sk commented:
Thanks for the updates - they look good. Only the ```statement_timeout``` is something that should get some more thoughts to ensure that ordinary queries will not get aborted (e.g. rendering of a big layer).

Regarding feature source / feature iterator, I have created ```QgsPostgresSharedData``` earlier for safe access to data that still needs to be shared between provider and iterators (of course, ideally there should be none). Transaction may be stored there. Still one needs to make sure that the transaction instance is not destroyed while being used.

wonder-sk commented:
> Yes, a transaction object can be shared between multiple instances and layers, so undo/redo with savepoints is a possibility. It is however difficult to guarantee that after doing undo and redo, the result is still the same as before.

Are you sure it wouldn't work? I would expect savepoints to behave reliably - at least for undo. Any example when it would fail?

mhugent commented:
>Are you sure it wouldn't work? I would expect savepoints to behave reliably - at least for undo. Any >example when it would fail?

I meant that the redo might be problematic, not the undo. E.g. after undo and redo, there might be a different state than before those two actions.

wonder-sk commented:
> I meant that the redo might be problematic, not the undo. E.g. after undo and redo, there might be a different state than before those two actions.

Sure, the state will be different because the sequence objects will generate different values. But will that necessarily lead to an inconsistent state? I'm just trying to understand if there are some issues inherent to the database software - or whether such issues are coming just from QGIS code (and if it would be difficult to fix them).

mhugent commented:
>I'm just trying to understand if there are some issues inherent to the database software

The issue is that postgres does not seem to have a redo function ( 'ROLLBACK TO SAVEPOINT implicitly destroys all savepoints that were established after the named savepoint' ). So no issue with QGIS code.

manisandro commented:
To solve the feature source / iterator issue, I added a `ref` method to `QgsPostgresConn` and increase/decrease it when creating/destroying a `QgsPostgresFeatureSource`. At the same time, I renamed `QgsPostgresConn::disconnect` to `QgsPostgresConn::unref`. Latter change is not strictly needed and I can revert it if desired. However, besides not being very inuitive, `disconnect` also collides with `QObject::disconnect` which is somewhat un-pretty.

Concerning `statement_timeout`: I'm tending towards exposing it in the `QgsTransaction` create methods. That way users are automatically aware of the issue and potential hangs of QGIS come less as a surprise. I guess a sane default would be around 10 seconds?

jef-n commented:
> The issue is that postgres does not seem to have a redo function ( 'ROLLBACK TO SAVEPOINT implicitly destroys all savepoints that were established after the named savepoint' ). So no issue with QGIS code.

Yes, we could only redo the last undone command - so undo should also take the previously undone command from the stack.

wonder-sk commented on src/providers/postgres/qgspostgresfeatureiterator.cpp:
```diff
@@ -555,8 +573,26 @@ QgsPostgresFeatureSource::QgsPostgresFeatureSource( const QgsPostgresProvider* p
     , mPrimaryKeyType( p->mPrimaryKeyType )
     , mPrimaryKeyAttrs( p->mPrimaryKeyAttrs )
     , mQuery( p->mQuery )
+    // static_cast since layers cannot be added to a transaction of a non-matching provider
```
obsolete comment?

wonder-sk commented:
> To solve the feature source / iterator issue, I added a ref method to QgsPostgresConn and increase/decrease it when creating/destroying a QgsPostgresFeatureSource.

That looks good to me. Maybe worth adding a comment about the solution (why the connection is stored in feature source).

> At the same time, I renamed QgsPostgresConn::disconnect to QgsPostgresConn::unref

Sounds reasonable.

> I guess a sane default would be around 10 seconds?

I have no idea...

wonder-sk commented on src/providers/postgres/CMakeLists.txt:
```diff
@@ -40,7 +44,7 @@ INCLUDE_DIRECTORIES(
 )
 
-ADD_LIBRARY (postgresprovider MODULE ${PG_SRCS} ${PG_MOC_SRCS})
+ADD_LIBRARY (postgresprovider MODULE ${PG_SRCS} ${PG_HDRS} ${PG_MOC_SRCS})
```
Why adding a header file to compilation?

manisandro replied:
Because it otherwise does not appear in the project file tree in various IDEs, QtCreator for one. It is common practice in CMake (and other build systems for the matter) to add headers also.

manisandro commented:
Ok I think all points are addressed. If that is the case, I'll rebase against latest master and clean up the commits.

mhugent commented:
Not sure why the travis build fails. From the 'details', it looks as if it didn't even start compiling. Maybe a timeout problem? @m-kuhn : any ideas?
